app/services/adjustment_create_service.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+class AdjustmentCreateService
+  include ServiceObjectErrorsMixin
+  attr_reader :adjustment
+
+  def initialize(adjustment_params)
+    @adjustment = Adjustment.new(adjustment_params)
+  end
+
+  def call
+    # Combine line items for adjustment
+    combine_adjustment
+    # Check for validity, and save the actual adjustment
+
+    if @adjustment.valid? && enough_inventory_for_decreases? && @adjustment.save
+      # Split into positive and negative portions
+      increasing_adjustment, decreasing_adjustment = @adjustment.split_difference
+      ActiveRecord::Base.transaction do
+        # Make the necessary changes in the db
+        @adjustment.save
+        @adjustment.storage_location.increase_inventory increasing_adjustment
+        @adjustment.storage_location.decrease_inventory decreasing_adjustment
+      rescue InsufficientAllotment => e
+        @adjustment.errors.add(:base, e.message)
+        raise InsufficientAllotment
```

Comment:
I'd just `raise e` or you'll lose the original backtrace.

Reply:
@dorner Done

Reply:
Sorry - this still is going to swallow the backtrace. I meant this:

```
rescue InsufficientAllotment => e
  @adjustment.errors.add(:base, e.message)
  raise e
```